### PR TITLE
Analysis workflow (#783)

### PR DESCRIPTION
**Pull request by G-D-Petrov**
#### What does this implement/fix? How does it work (high level)? Highlight notable design decisions.
Implements a workflow for code coverage that will be extended to support other analysis tasks
Addresses issue #638

### CODE REVIEW

File: .github/workflows/analysis_workflow.yml [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+          echo "ARCTICDB_REAL_S3_SECRET_KEY=${{ secrets.AWS_S3_SECRET_KEY }}" >> $GITHUB_ENV
+
+      - name: Test with pytest
+        shell: bash -l {0}
+        run: |
+          yum install nodejs npm -y
+          npm config set strict-ssl false
+          npm install -g azurite
+          /opt/python/cp36-cp36m/bin/python -m pip install --upgrade pip
+          /opt/python/cp36-cp36m/bin/python -m pip install pycparser six cffi urllib3 python-dateutil PyJWT jmespath idna cryptography charset-normalizer certifi requests py pluggy path oauthlib MarkupSafe iniconfig dataclasses botocore attrs Werkzeug termcolor s3transfer requests-oauthlib pytz pytest protobuf portalocker path.py numpy msal mock Jinja2 itsdangerous isodate grpcio execnet decorator contextlib2 click azure-core xxhash xmltodict sortedcontainers retry responses pyyaml pytest-shutil pytest-fixture-config psutil prometheus-client pandas msrest msgpack msal-extensions grpcio-tools future flask colorama boto3 pytest-timeout pytest-server-fixtures pytest-cpp pymongo moto hypothesis flask-cors azure-storage-blob azure-identity coverage pytest-xdist
```
Comment (poodlewars) [inline .github/workflows/analysis_workflow.yml:190]: This won't be maintainable when deps change. We have a similar problem elsewhere (shared an internal link with you). Guess we can do it that way.

File: .github/workflows/analysis_workflow.yml [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+          echo "ARCTICDB_REAL_S3_ACCESS_KEY=${{ secrets.AWS_S3_ACCESS_KEY }}" >> $GITHUB_ENV
+          echo "ARCTICDB_REAL_S3_SECRET_KEY=${{ secrets.AWS_S3_SECRET_KEY }}" >> $GITHUB_ENV
+
+      - name: Test with pytest
+        shell: bash -l {0}
+        run: |
+          yum install nodejs npm -y
+          npm config set strict-ssl false
+          npm install -g azurite
+          /opt/python/cp36-cp36m/bin/python -m pip install --upgrade pip
```
Comment (poodlewars) [inline .github/workflows/analysis_workflow.yml:189]: Shall we use Python 3.11? 3.6 is end of life and we will probably remove support for it once users have all moved away from it. So using 3.11 will be safer for longer.

File: .github/workflows/analysis_workflow.yml [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                  "type": "section",
+                  "text": {
+                    "type": "mrkdwn",
+                    "text": "The CPP Code Coverage from the current run(${{ env.CPP_COV_PERCENT }}%) is lower the previous one(${{ env.CPP_COV_PREV_PERCENT }}%)."
+                  }
+                }
+              ]
+            }
+        env:
+          SLACK_WEBHOOK_URL: ${{ secrets.TEST_WEBHOOK_URL }}
```
Comment (poodlewars) [inline .github/workflows/analysis_workflow.yml:283]: Change to the non-test URL

File: .github/workflows/analysis_workflow.yml [hunk 4]
```diff
@@ -0,0 +1,293 @@
+name: Build with analysis tools
+on:
+  workflow_dispatch:
+    inputs:
+      run:
+        type: boolean
+
+  push:
```
[the remark this answers is not in the document]
Comment (G-D-Petrov) [reply .github/workflows/analysis_workflow.yml:8]: Yes, this is just for testing, I will changing before merging.

File: .github/workflows/analysis_workflow.yml [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+      - build_tooling/**
+      - cpp/**
+      - python/**
+      - setup.*
+      - pyproject.toml
+jobs:
+  common_config:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
```
[the remark this answers is not in the document]
Comment (G-D-Petrov) [reply .github/workflows/analysis_workflow.yml:22]: This matrix part is only for the seed jobs, so the persistent tests work correctly.
I will look into how to do this better.

File: python/tests/stress/arcticdb/version_store/test_stress_version_map_compact.py
Comment (poodlewars) [inline python/tests/stress/arcticdb/version_store/test_stress_version_map_compact.py:1]: Please could you explain why we needed to change this test?

I'm also confused by the `error` logic. You have `while (not done.value) or error.value:`, but why would we want to keep going after an error?

Also, why did the `idx2` range need to change?

Comment (poodlewars) [reply python/tests/stress/arcticdb/version_store/test_stress_version_map_compact.py:1]: This is from William's stuff. I've asked this same question on https://github.com/man-group/ArcticDB/pull/690/files

File: .github/workflows/analysis_workflow.yml [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+
+      - name: Get number of CPU cores
+        uses: SimenB/github-actions-cpu-cores@v1
+        id: cpu-cores
+
+      - name: Configure sccache
```
Comment (poodlewars) [inline .github/workflows/analysis_workflow.yml:135]: Probably don't want this (we'll just invalidate the cache for the normal builds)

File: .github/workflows/analysis_workflow.yml [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+    env:
+      SCCACHE_GHA_VERSION: ${{vars.SCCACHE_GHA_VERSION || 1}} # Setting this env var enables the caching
+      VCPKG_NUGET_USER: ${{secrets.VCPKG_NUGET_USER || github.repository_owner}}
+      VCPKG_NUGET_TOKEN: ${{secrets.VCPKG_NUGET_TOKEN || secrets.GITHUB_TOKEN}}
+      VCPKG_MAN_NUGET_USER: ${{secrets.VCPKG_MAN_NUGET_USER}} # For forks to download pre-compiled dependencies from the Man repo
+      VCPKG_MAN_NUGET_TOKEN: ${{secrets.VCPKG_MAN_NUGET_TOKEN}}
+      CMAKE_C_COMPILER_LAUNCHER: sccache
+      CMAKE_CXX_COMPILER_LAUNCHER: sccache
+      ARCTIC_CMAKE_PRESET: linux-debug
+      CIBW_ENVIRONMENT_PASS_LINUX: SCCACHE_GHA_VERSION ACTIONS_CACHE_URL ACTIONS_RUNTIME_TOKEN VCPKG_INSTALLATION_ROOT
```
Comment (poodlewars) [inline .github/workflows/analysis_workflow.yml:123]: Why do you need CIBW stuff in this file?

File: .github/workflows/analysis_workflow.yml [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+          name: cpp-coverage-artifact
+          path: cpp/out/linux-debug-build/coverage.zip
+
+      - name: Upload Python Coverage
+        uses: actions/upload-artifact@v3
+        with:
+          name: python-coverage-artifact
+          path: python/python_cov.zip
+
+      - name: Restore cpp cache
```
Comment (poodlewars) [inline .github/workflows/analysis_workflow.yml:146]: Name wrong?

**Review by poodlewars — APPROVED** [review]

**Comment by poodlewars** [comment]
Please remember to squash the commits

**Pull request merged**